File: sources/r/10_pi.R
# task 10 pi — expected output: 44889
# build: Rscript 10_pi.R    run: Rscript 10_pi.R
#
# Unbounded spigot (Gibbons), emitting the leading 3 first, summing the first
# 10000 digits it emits. R has no arbitrary-precision integers, so the state
# (q, r, t) is carried as sign/magnitude vectors of base-10^9 limbs, with the
# multiply-by-a-small-integer, the add, the subtract and the compare written
# by hand. Every intermediate stays below 2^53, so doubles hold limbs exactly.
# Both divisions in the spigot have a quotient below 100, so the big/big
# division is a normalised two-limb estimate plus a correction step or two.

BASE <- 1e9
NDIGITS <- 10000

# --- base-10^9 limbs, little-endian, no leading zero limbs -------------------

mk <- function(m, s) {
  n <- length(m)
  while (n > 1 && m[n] == 0) n <- n - 1
  if (n == 1 && m[1] == 0) return(list(s = 1, m = 0))
  if (n < length(m)) m <- m[seq_len(n)]
  list(s = s, m = m)
}

ZERO <- list(s = 1, m = 0)
ONE <- list(s = 1, m = 1)

cmp_mag <- function(a, b) {
  if (length(a) != length(b)) return(if (length(a) > length(b)) 1 else -1)
  d <- which(a != b)
  if (length(d) == 0) return(0)
  i <- d[length(d)]
  if (a[i] > b[i]) 1 else -1
}

cmp <- function(x, y) {
  if (x$s != y$s) return(if (x$s > y$s) 1 else -1)
  x$s * cmp_mag(x$m, y$m)
}

add_mag <- function(a, b) {
  la <- length(a)
  lb <- length(b)
  n <- if (la > lb) la else lb
  s <- numeric(n)
  s[seq_len(la)] <- a
  s[seq_len(lb)] <- s[seq_len(lb)] + b
  s <- c(s, 0)
  repeat {
    o <- s >= BASE
    if (!any(o)) break
    o <- as.numeric(o)
    s <- s - o * BASE
    s <- s + c(0, o[seq_len(length(s) - 1)])
  }
  n <- length(s)
  while (n > 1 && s[n] == 0) n <- n - 1
  if (n < length(s)) s <- s[seq_len(n)]
  s
}

sub_mag <- function(a, b) {          # requires a >= b
  s <- a
  lb <- length(b)
  s[seq_len(lb)] <- s[seq_len(lb)] - b
  repeat {
    o <- s < 0
    if (!any(o)) break
    o <- as.numeric(o)
    s <- s + o * BASE
    s <- s - c(0, o[seq_len(length(s) - 1)])
  }
  n <- length(s)
  while (n > 1 && s[n] == 0) n <- n - 1
  if (n < length(s)) s <- s[seq_len(n)]
  s
}

add <- function(x, y) {
  if (x$s == y$s) return(mk(add_mag(x$m, y$m), x$s))
  k <- cmp_mag(x$m, y$m)
  if (k == 0) return(ZERO)
  if (k > 0) mk(sub_mag(x$m, y$m), x$s) else mk(sub_mag(y$m, x$m), y$s)
}

sub <- function(x, y) {
  if (length(y$m) == 1 && y$m[1] == 0) return(x)
  add(x, list(s = -y$s, m = y$m))
}

mul_small <- function(x, m) {
  if (m == 0) return(ZERO)
  p <- x$m * m
  lo <- p %% BASE
  hi <- (p - lo) / BASE
  s <- c(lo, 0) + c(0, hi)
  repeat {
    o <- s >= BASE
    if (!any(o)) break
    o <- as.numeric(o)
    s <- s - o * BASE
    s <- s + c(0, o[seq_len(length(s) - 1)])
  }
  mk(s, x$s)
}

# floor(A / B) for magnitudes. The spigot's quotients are below 100 and the
# estimate from the two leading limbs is accurate to within one, so the
# correction loops take a step or two; they are exact for any quotient.
div_quot <- function(A, B) {
  if (cmp_mag(A, B) < 0) return(0)
  la <- length(A)
  lb <- length(B)
  a0 <- if (la >= 2) A[la - 1] else 0
  b0 <- if (lb >= 2) B[lb - 1] else 0
  est <- ((A[la] + a0 / BASE) / (B[lb] + b0 / BASE)) * BASE^(la - lb)
  q <- if (is.finite(est) && est > 0) floor(est) else 0
  if (q > 4096) q <- 4096
  P <- mul_small(list(s = 1, m = B), q)$m
  while (cmp_mag(P, A) > 0) {
    q <- q - 1
    P <- sub_mag(P, B)
  }
  repeat {
    Q <- add_mag(P, B)
    if (cmp_mag(Q, A) > 0) break
    P <- Q
    q <- q + 1
  }
  q
}

# --- the spigot --------------------------------------------------------------

pi_digit_sum <- function() {
  q <- ONE
  r <- ZERO
  t <- ONE
  k <- 1
  n <- 3
  l <- 3
  count <- 0
  total <- 0
  while (count < NDIGITS) {
    if (cmp(sub(add(mul_small(q, 4), r), t), mul_small(t, n)) < 0) {
      total <- total + n
      count <- count + 1
      nxt <- div_quot(mul_small(add(mul_small(q, 3), r), 10)$m, t$m) - 10 * n
      newr <- mul_small(sub(r, mul_small(t, n)), 10)
      q <- mul_small(q, 10)
      r <- newr
      n <- nxt
    } else {
      nxt <- div_quot(add(mul_small(q, 7 * k + 2), mul_small(r, l))$m,
                      mul_small(t, l)$m)
      newr <- mul_small(add(mul_small(q, 2), r), l)
      newt <- mul_small(t, l)
      q <- mul_small(q, k)
      r <- newr
      t <- newt
      k <- k + 1
      n <- nxt
      l <- l + 2
    }
  }
  cat(sprintf("%.0f\n", total))
}

pi_digit_sum()

File: sources/swift/10_pi.swift
// task 10 pi — expected output: 44889
// build: swiftc -O -o prog 10_pi.swift    run: ./prog

// Gibbons' unbounded spigot, from "Unbounded Spigot Algorithms for the Digits of Pi":
//
//     pi = g(1,0,1,1,3,3) where
//         g(q,r,t,k,n,l) = if 4*q+r-t < n*t
//                          then n : g(10*q, 10*(r-n*t), t, k,
//                                     div (10*(3*q+r)) t - 10*n, l)
//                          else g(q*k, (2*q+r)*l, t*l, k+1,
//                                     div (q*(7*k+2) + r*l) (t*l), l+2)
//
// The Swift standard library has no arbitrary-precision integers, so the state is
// carried in hand-written ones: sign-magnitude, base 10^9 limbs, with
// multiply-by-small and a division whose quotient is known to be small. Only the
// sum of the digits is printed, never the digits themselves.

let BASE: UInt64 = 1_000_000_000

struct Big {
    // Little-endian base-10^9 limbs, normalised: no leading zero limbs.
    var d: [UInt64] = [0]

    init() {}

    init(_ v: UInt64) {
        var x = v
        var limbs: [UInt64] = []
        while x > 0 {
            limbs.append(x % BASE)
            x /= BASE
        }
        d = limbs.isEmpty ? [0] : limbs
    }

    var isZero: Bool { return d.count == 1 && d[0] == 0 }

    mutating func trim() {
        while d.count > 1 && d[d.count - 1] == 0 {
            d.removeLast()
        }
    }

    func cmp(_ other: Big) -> Int {
        if d.count != other.d.count {
            return d.count < other.d.count ? -1 : 1
        }
        var i = d.count - 1
        while i >= 0 {
            if d[i] != other.d[i] {
                return d[i] < other.d[i] ? -1 : 1
            }
            i -= 1
        }
        return 0
    }

    func add(_ other: Big) -> Big {
        let n = d.count > other.d.count ? d.count : other.d.count
        var res = [UInt64](repeating: 0, count: n + 1)
        var carry: UInt64 = 0
        for i in 0..<n {
            let x = (i < d.count ? d[i] : 0) + (i < other.d.count ? other.d[i] : 0) + carry
            res[i] = x % BASE
            carry = x / BASE
        }
        res[n] = carry
        var r = Big()
        r.d = res
        r.trim()
        return r
    }

    // Requires self >= other.
    func sub(_ other: Big) -> Big {
        var res = d
        var borrow: UInt64 = 0
        for i in 0..<res.count {
            let b = i < other.d.count ? other.d[i] : 0
            let x = res[i]
            if x >= b + borrow {
                res[i] = x - b - borrow
                borrow = 0
            } else {
                res[i] = x + BASE - b - borrow
                borrow = 1
            }
        }
        var r = Big()
        r.d = res
        r.trim()
        return r
    }

    // Requires m < 10^9 so that a limb times m plus the carry still fits in 64 bits.
    func mulSmall(_ m: UInt64) -> Big {
        if m == 0 || isZero {
            return Big(0)
        }
        var res = [UInt64](repeating: 0, count: d.count + 1)
        var carry: UInt64 = 0
        for i in 0..<d.count {
            let p = d[i] * m + carry
            res[i] = p % BASE
            carry = p / BASE
        }
        res[d.count] = carry
        var r = Big()
        r.d = res
        r.trim()
        return r
    }

    // Only used on values known to be below 10^18.
    func toUInt64() -> UInt64 {
        var v: UInt64 = 0
        var i = d.count - 1
        while i >= 0 {
            v = v * BASE + d[i]
            i -= 1
        }
        return v
    }

    // floor(self / other). The spigot only ever asks for a digit-extraction
    // quotient, a few hundred at most, so the two-limb estimate below is within a
    // couple of units of the truth and the correction loops run a few times.
    func divMod(_ other: Big) -> (Big, Big) {
        if cmp(other) < 0 {
            return (Big(0), self)
        }
        let la = d.count
        let lb = other.d.count
        var q: UInt64
        if la <= 2 && lb <= 2 {
            q = toUInt64() / other.toUInt64()
        } else {
            let secondA: UInt64 = la >= 2 ? d[la - 2] : 0
            let secondB: UInt64 = lb >= 2 ? other.d[lb - 2] : 0
            let topA = Double(d[la - 1]) * 1e9 + Double(secondA)
            let topB = Double(other.d[lb - 1]) * 1e9 + Double(secondB)
            var scale = 1.0
            var e = la - lb
            while e > 0 {
                scale *= 1e9
                e -= 1
            }
            var est = (topA / topB) * scale
            if est > 1e9 {
                est = 1e9
            }
            if est < 1.0 {
                est = 1.0
            }
            q = UInt64(est)
        }
        while other.mulSmall(q).cmp(self) > 0 {
            q -= 1
        }
        while other.mulSmall(q + 1).cmp(self) <= 0 {
            q += 1
        }
        return (Big(q), sub(other.mulSmall(q)))
    }
}

// Signed wrapper: the spigot's r goes negative, everything else stays positive.
struct Signed {
    var neg: Bool
    var mag: Big

    init(_ m: Big, neg: Bool = false) {
        if m.isZero {
            self.neg = false
            self.mag = m
        } else {
            self.neg = neg
            self.mag = m
        }
    }

    static let zero = Signed(Big(0))

    var isZero: Bool { return mag.isZero }

    func negated() -> Signed {
        return isZero ? self : Signed(mag, neg: !neg)
    }

    func add(_ other: Signed) -> Signed {
        if neg == other.neg {
            return Signed(mag.add(other.mag), neg: neg)
        }
        let c = mag.cmp(other.mag)
        if c == 0 {
            return Signed.zero
        }
        if c > 0 {
            return Signed(mag.sub(other.mag), neg: neg)
        }
        return Signed(other.mag.sub(mag), neg: other.neg)
    }

    func sub(_ other: Signed) -> Signed {
        return add(other.negated())
    }

    func mulSmall(_ m: UInt64) -> Signed {
        return Signed(mag.mulSmall(m), neg: neg)
    }

    func cmp(_ other: Signed) -> Int {
        if neg != other.neg {
            return neg ? -1 : 1
        }
        let c = mag.cmp(other.mag)
        return neg ? -c : c
    }
}

var q = Big(1)
var r = Signed(Big(0))
var t = Big(1)
var n: UInt64 = 3
var k: UInt64 = 1
var l: UInt64 = 3
var digitSum = 0
var produced = 0

while produced < 10000 {
    // 4q + r - t < n t is the same test as 4q + r < (n + 1) t, with both sides
    // non-negative.
    let lhs = Signed(q.mulSmall(4)).add(r)
    let rhs = Signed(t.mulSmall(n + 1))

    if lhs.cmp(rhs) < 0 {
        // The digit n is settled: emit it and scale the state by ten.
        digitSum += Int(n)
        produced += 1

        let threeQR = Signed(q.mulSmall(3)).add(r)                  // 3q + r, never negative
        let (quot, _) = threeQR.mag.mulSmall(10).divMod(t)           // div (10 (3q + r)) t
        let next = quot.toUInt64() - 10 * n

        let nq = q.mulSmall(10)
        let nr = r.sub(Signed(t.mulSmall(n))).mulSmall(10)           // 10 (r - n t)
        q = nq
        r = nr
        n = next
    } else {
        // Not settled yet: widen the state by one more term of the series.
        let num = Signed(q.mulSmall(7 * k + 2)).add(r.mulSmall(l))   // q (7k + 2) + r l
        let nt = t.mulSmall(l)
        let (quot, _) = num.mag.divMod(nt)
        let next = quot.toUInt64()

        let nq = q.mulSmall(k)
        let nr = Signed(q.mulSmall(2)).add(r).mulSmall(l)            // (2q + r) l
        q = nq
        r = nr
        t = nt
        n = next
        k += 1
        l += 2
    }
}

print(digitSum)
